plutus-ledger-api/test-plugin/Spec/ReturnUnit/V1.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE BangPatterns          #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NegativeLiterals      #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PatternSynonyms       #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE ViewPatterns          #-}
+{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:target-version=1.0.0 #-}
+
+module Spec.ReturnUnit.V1 where
+
+import PlutusLedgerApi.Common.Versions
+import PlutusLedgerApi.Test.V1.EvaluationContext qualified as V1
+import PlutusLedgerApi.V1 as V1
+import PlutusPrelude
+import PlutusTx.Builtins qualified as PlutusTx
+import PlutusTx.Code
+import PlutusTx.IsData qualified as PlutusTx
+import PlutusTx.Prelude (BuiltinUnit, check)
+import PlutusTx.TH (compile)
+
+import Test.Tasty (TestName, TestTree, testGroup)
+import Test.Tasty.HUnit
+
+import Control.Monad.Writer
+
+tests :: TestTree
+tests =
+  testGroup
+    "Plutus V1 validators may evaluate to any value"
+    [ expectSuccess "should succeed" good (I 1)
+    , expectSuccess "returns Bool" returnsBool (I 1)
+    , expectSuccess "too many parameters" tooManyParameters (I 1)
+    ]
+
+evalCtx :: V1.EvaluationContext
+evalCtx =
+  fst . unsafeFromRight . runWriterT . V1.mkEvaluationContext $
+    fmap snd V1.costModelParamsForTesting
+
+expectSuccess ::
+  forall a.
+  TestName ->
+  CompiledCode (BuiltinData -> a) ->
+  -- | Script argument
+  Data ->
+  TestTree
+expectSuccess name code arg = testCase name $ case res of
+  Left _  -> assertFailure "fails"
+  Right _ -> pure ()
+  where
+    sScript = serialiseCompiledCode code
+    script = either (error . show) id $ V1.deserialiseScript conwayPV sScript
```

Comment:
Nit: is this serialization-deserialization roundtrip necessary?

Reply:
Yes, `ScriptForEvaluation` can only be obtained via `deserialiseScript`, which is by design.

Reply:
Thanks for the explanation!